apollo-router/src/plugins/file_uploads/rearange_query_plan.rs

```diff
@@ -0,0 +1,244 @@
+use std::cmp;
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::collections::HashSet;
+
+use indexmap::IndexSet;
+use itertools::Itertools;
+
+use super::error::FileUploadError;
+use super::MapPerVariable;
+use super::UploadResult;
+use crate::query_planner::DeferredNode;
+use crate::query_planner::FlattenNode;
+use crate::query_planner::PlanNode;
+use crate::services::execution::QueryPlan;
+
+pub(super) fn rearange_query_plan(
+    query_plan: &QueryPlan,
+    files_order: &IndexSet<String>,
+    map_per_variable: &MapPerVariable,
+) -> UploadResult<QueryPlan> {
+    let root = &query_plan.root;
+    let mut variable_ranges = HashMap::new();
+    for (name, map) in map_per_variable.iter() {
+        variable_ranges.insert(
+            name.as_str(),
+            map.keys()
+                .map(|file| files_order.get_index_of(file))
+                .minmax()
+                .into_option()
+                .expect("map always have keys"),
+        );
+    }
+
+    let root = rearrange_plan_node(root, &mut HashMap::new(), &variable_ranges)?;
+    Ok(QueryPlan {
+        root,
+        usage_reporting: query_plan.usage_reporting.clone(),
+        formatted_query_plan: query_plan.formatted_query_plan.clone(),
+        query: query_plan.query.clone(),
+    })
+}
+
+// Recursive, and recursion is safe here since query plan is executed recursively.
+fn rearrange_plan_node<'a>(
```

Comment:
This function needs more documentation I think to explain what's going on inside each match arms

Reply:
Makes sense algorithm is pretty complex 👍 
I added docs.